zenith_utils/src/sync.rs

```diff
@@ -0,0 +1,133 @@
+use pin_project_lite::pin_project;
+use std::future::Future;
+use std::pin::Pin;
+use std::{io, task};
+
+pin_project! {
+    /// We use this future to mark certain methods
+    /// as callable in both sync and async modes
+    #[repr(transparent)]
+    pub struct SyncFuture<S, T: Future> {
+        #[pin]
+        nested: T,
+        _marker: std::marker::PhantomData<S>,
+    }
+}
+
+impl<S, T: Future> SyncFuture<S, T> {
+    pub(crate) fn new(nested: T) -> Self {
+        Self {
+            nested,
+            _marker: Default::default(),
```

Comment:
NIT:

Phantom data could be just constructed, without trait solving magic involved:  `_marker: PhantomData,`

Reply:
Sure, why not.